File: P2.java
// Time Complexity :    O(m x n)
// Space Complexity :   O(m x n)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english - Moving up and down basically requires moving of the pointers which starts with the direction condition which if true will move up or else will move down.

// Your code here along with comments explaining your approach

class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;
        int[] result = new int[m * n];
        int idx = 0;

        if (mat == null) return result;
        int i = 0; int j = 0;       //matrix location
        boolean dir = true;
        while (idx < m * n) {    
            result[idx] = mat[i][j];    //initialising index with 0,0 positions
            idx ++;
            //find new location of i, j
            if (dir) {
                //up
                if (j == n - 1) {
                    i ++;
                    dir = false;
                } else if (i == 0) {    //this is second condition to avoid out of bounds error at top right corner
                    j ++;
                    dir = false;
                } else {
                    i --;
                    j ++;
                }
            } else {
                //moving down
                if (i == m - 1) {
                    j ++;
                    dir = true;
                } else if (j == 0) {    //same reason why this is the second condition at bottom left corner
                    i ++;
                    dir = true;
                } else {
                    i ++; 
                    j --;
                }
            }
        }
        return result;
    }
}

File: p2.py
#Time complexity o(mn)
#Space complexity o(1)
#Worked on Leet code:yes

class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        if not mat or not mat[0]:
            return []
        r = len(mat)
        c= len(mat[0])
        res= [None]*(c*r)
        i=0
        matr=0
        matc=0
        di=1 #1=up,0=down

        while i<(r*c):
            res[i]= mat[matr][matc]
            if di ==1: #moving up
                if matc+1>=c: #right boundry
                    matr+=1
                    di=0
                elif matr-1<0: # Top Boundry
                    matc+=1
                    di=0
                else:
                    matr-=1
                    matc+=1
            else:  #moving down
                if matr+1>=r: # bottom boundry
                    matc+=1
                    di=1
                elif matc-1<0: #  Left boundry
                    matr+=1
                    di=1
                else:
                    matr+=1
                    matc-=1
            i+=1
        return res


